tkg/client/management_components.go

```diff
@@ -406,38 +406,38 @@ func GetKappControllerConfigValuesFile(userConfigValuesFile, kappControllerValue
 // file to provide a source of keys to filter for the valid user provided values.
 // For example, this function uses config_default.yaml filepath to find relevant config variables
 // and returns the config map of user provided variable among all applicable config variables
-func (c *TkgClient) GetUserConfigVariableValueMap(configDefaultFilePath string, rw tkgconfigreaderwriter.TKGConfigReaderWriter) (map[string]interface{}, error) {
+func (c *TkgClient) GetUserConfigVariableValueMap(configDefaultFilePath string, rw tkgconfigreaderwriter.TKGConfigReaderWriter, withDefaults bool) (map[string]interface{}, error) {
 	bytes, err := os.ReadFile(configDefaultFilePath)
 	if err != nil {
 		return nil, err
 	}
 
-	variables, err := GetConfigVariableListFromYamlData(bytes)
+	variablesMap, err := GetConfigVariableListFromYamlData(bytes)
 	if err != nil {
 		return nil, err
 	}
 
 	userProvidedConfigValues := map[string]interface{}{}
-	for _, k := range variables {
-		if v, e := rw.Get(k); e == nil {
-			userProvidedConfigValues[k] = utils.Convert(v)
+	for key, value := range variablesMap {
+		v, e := rw.Get(key)
+		if e == nil {
+			userProvidedConfigValues[key] = utils.Convert(v)
+		} else {
+			if withDefaults {
```

Comment:
Can you explain which variables are missing? I am not sure Why TKGConfigReaderWriter does not have them? cc @tenczar @wjun @12345lcr

Reply:
From my experiment whatever user has customized is only there in tkgconfigreaderwriter.TKGConfigReaderWriter. The default values from config_default.yaml was missing. 

I can do one more experiment and let you know what variable was missing.

Reply:
I printed all the key value pairs under else and this was printed:

```
Key ANTREA_NO_SNAT: Value %!s(bool=false) not present
Key VSPHERE_CONTROL_PLANE_ENDPOINT_PORT: Value %!s(int=6443) not present
Key AWS_SECURITY_GROUP_NODE: Value  not present
Key STORAGE_CLASSES: Value %!s(<nil>) not present
Key AZURE_VNET_NAME: Value  not present
Key AZURE_NODE_OUTBOUND_LB_FRONTEND_IP_COUNT: Value %!s(int=1) not present
Key SUPERVISOR_ISSUER_CA_BUNDLE_DATA_B64: Value %!s(<nil>) not present
Key AVI_DISABLE_INGRESS_CLASS: Value %!s(bool=true) not present
Key ANTREA_MULTICAST_INTERFACES: Value  not present
Key TKR_DATA: Value %!s(<nil>) not present
Key WORKER_ROLLOUT_STRATEGY: Value RollingUpdate not present
Key AUTOSCALER_MIN_SIZE_0: Value %!s(<nil>) not present
Key SERVICE_DOMAIN: Value %!s(<nil>) not present
Key SUPERVISOR_ISSUER_URL: Value %!s(<nil>) not present
Key OCI_CONTROL_PLANE_PV_TRANSIT_ENCRYPTION: Value %!s(bool=true) not present
Key ANTREA_ENABLE_USAGE_REPORTING: Value %!s(bool=false) not present
Key NSXT_CLIENT_CERT_DATA: Value  not present
Key AZURE_IMAGE_GALLERY: Value %!s(<nil>) not present
Key ENABLE_MHC_WORKER_NODE: Value %!s(bool=true) not present
Key AVI_NAMESPACE: Value tkg-system-networking not present
Key AVI_SERVICE_ENGINE_GROUP: Value %!s(<nil>) not present
Key VSPHERE_MEM_MIB: Value %!s(int=4096) not present
Key VSPHERE_CONTROL_PLANE_NUM_CPUS: Value %!s(int=2) not present
Key AWS_SECURITY_GROUP_BASTION: Value  not present
Key AWS_CONTROL_PLANE_OS_DISK_SIZE_GIB: Value %!s(int=80) not present
Key TKG_IP_FAMILY: Value %!s(<nil>) not present
Key ANTREA_MULTICAST_IGMPQUERY_INTERVAL: Value 125s not present
Key ANTREA_DISABLE_TXCHECKSUM_OFFLOAD: Value %!s(bool=false) not present
Key OCI_REGION: Value us-phoenix-1 not present
Key VSPHERE_TLS_THUMBPRINT: Value  not present
Key AVI_INGRESS_NODE_NETWORK_LIST: Value  not present
Key ANTREA_NODEPORTLOCAL: Value %!s(bool=true) not present
Key ANTREA_TRAFFIC_CONTROL: Value %!s(bool=false) not present
Key OCI_SSH_KEY: Value %!s(<nil>) not present
Key AZURE_NODE_SUBNET_CIDR: Value 10.0.1.0/24 not present
Key AUTOSCALER_MIN_SIZE_2: Value %!s(<nil>) not present
Key AVI_ENABLE: Value %!s(bool=false) not present
Key AVI_AKO_IMAGE_PULL_POLICY: Value IfNotPresent not present
Key ANTREA_MULTICLUSTER: Value %!s(bool=false) not present
Key TKR_DISCOVER_FREQUENCY: Value %!s(int=600) not present
Key NODE_POOL_0_TAINTS: Value %!s(<nil>) not present
Key AZURE_CONTROL_PLANE_OS_DISK_SIZE_GIB: Value %!s(int=128) not present
Key AVI_CONTROL_PLANE_HA_PROVIDER: Value %!s(bool=false) not present
Key AVI_LABELS: Value %!s(<nil>) not present
Key ANTREA_MULTICLUSTER_NAMESPACE: Value  not present
Key WORKER_MACHINE_COUNT_1: Value %!s(<nil>) not present
Key VSPHERE_STORAGE_POLICY_ID: Value  not present
Key AZURE_IDENTITY_NAME: Value %!s(<nil>) not present
Key WORKER_NODE_SEARCH_DOMAINS: Value %!s(<nil>) not present
Key REMOVE_CRS_FOR_ADDON_TYPE: Value %!s(<nil>) not present
Key OCI_EXTERNAL_WORKER_SUBNET_ID: Value %!s(<nil>) not present
Key DISABLE_TMC_CLOUD_PERMISSIONS: Value %!s(bool=false) not present
Key CERT_RENEW_BEFORE: Value 360h not present
Key ANTREA_PROXY_ALL: Value %!s(bool=false) not present
Key ANTREA_FLOWEXPORTER: Value %!s(bool=false) not present
Key OCI_NODE_MACHINE_TYPE: Value VM.Standard.E4.Flex not present
Key ANTREA_ENDPOINTSLICE: Value %!s(bool=true) not present
Key VSPHERE_IGNORE_PCI_DEVICES_ALLOW_LIST: Value %!s(<nil>) not present
Key NSXT_CLIENT_CERT_KEY_DATA: Value  not present
Key DEFAULT_STORAGE_CLASS: Value %!s(<nil>) not present
Key AZURE_CONTROL_PLANE_MACHINE_TYPE: Value Standard_D2s_v3 not present
Key AZURE_VNET_RESOURCE_GROUP: Value  not present
Key AUTOSCALER_MAX_NODE_PROVISION_TIME: Value 15m not present
Key VSPHERE_DISK_GIB: Value %!s(int=40) not present
Key VSPHERE_WORKER_PCI_DEVICES: Value %!s(<nil>) not present
Key AZURE_IMAGE_THIRD_PARTY: Value %!s(<nil>) not present
Key ANTREA_TRAFFIC_ENCRYPTION_MODE: Value none not present
Key NSXT_MANAGER_HOST: Value  not present
Key NODE_MACHINE_TYPE_1: Value  not present
Key OIDC_GROUPS_CLAIM: Value %!s(<nil>) not present
Key USE_TOPOLOGY_CATEGORIES: Value %!s(bool=false) not present
Key ENABLE_AUTOSCALER: Value %!s(bool=false) not present
Key AWS_IDENTITY_REF_KIND: Value  not present
Key AVI_USERNAME: Value  not present
Key AVI_CONTROL_PLANE_NETWORK_CIDR: Value %!s(<nil>) not present
Key AVI_DISABLE_STATIC_ROUTE_SYNC: Value %!s(bool=true) not present
Key AZURE_NODE_MACHINE_TYPE: Value Standard_D2s_v3 not present
Key AUTOSCALER_SCALE_DOWN_UNNEEDED_TIME: Value 10m not present
Key AVI_INGRESS_DEFAULT_INGRESS_CONTROLLER: Value %!s(bool=false) not present
Key VSPHERE_DATASTORE: Value %!s(<nil>) not present
Key TKG_PROXY_CA_CERT: Value  not present
Key AVI_DATA_NETWORK_CIDR: Value %!s(<nil>) not present
Key AZURE_ENVIRONMENT: Value AzurePublicCloud not present
Key TKG_CUSTOM_IMAGE_REPOSITORY_SKIP_TLS_VERIFY: Value %!s(bool=false) not present
Key CONTROL_PLANE_NODE_NAMESERVERS: Value %!s(<nil>) not present
Key AVI_INGRESS_SERVICE_TYPE: Value  not present
Key VSPHERE_AZ_2: Value %!s(<nil>) not present
Key VSPHERE_INSECURE: Value %!s(bool=false) not present
Key WORKER_VM_CLASS: Value %!s(<nil>) not present
Key AZURE_NODE_AZ_1: Value  not present
Key ANTREA_SECONDARY_NETWORK: Value %!s(bool=false) not present
Key SIZE: Value %!s(<nil>) not present
Key VSPHERE_NETWORK: Value VM Network not present
Key LDAP_USER_SEARCH_ID_ATTRIBUTE: Value DN not present
Key ANTREA_FLOWEXPORTER_IDLE_TIMEOUT: Value 15s not present
Key OCI_NODE_PV_TRANSIT_ENCRYPTION: Value %!s(bool=true) not present
Key VSPHERE_FOLDER: Value %!s(<nil>) not present
Key NSXT_PASSWORD: Value  not present
Key AZURE_NODE_DATA_DISK_SIZE_GIB: Value %!s(int=256) not present
Key AVI_MANAGEMENT_CLUSTER_CONTROL_PLANE_VIP_NETWORK_NAME: Value %!s(<nil>) not present
Key ANTREA_PROXY_SKIP_SERVICES: Value  not present
Key VSPHERE_CONTROL_PLANE_CUSTOM_VMX_KEYS: Value %!s(<nil>) not present
Key NSXT_VMC_ACCESS_TOKEN: Value  not present
Key AVI_CA_DATA_B64: Value  not present
Key ANTREA_POLICY: Value %!s(bool=true) not present
Key DISABLE_CRS_FOR_ADDON_TYPE: Value %!s(<nil>) not present
Key ENABLE_TKGS_ON_VSPHERE7: Value %!s(<nil>) not present
Key VSPHERE_AZ_0: Value %!s(<nil>) not present
Key AWS_NODE_OS_DISK_SIZE_GIB: Value %!s(int=80) not present
Key AVI_CONTROLLER_VERSION: Value 20.1.3 not present
Key ANTREA_TRANSPORT_INTERFACE: Value %!s(<nil>) not present
Key OCI_CONTROL_PLANE_MACHINE_TYPE: Value VM.Standard.E4.Flex not present
Key OCI_EXTERNAL_CONTROL_PLANE_EP_SUBNET_ID: Value %!s(<nil>) not present
Key VSPHERE_CONTROL_PLANE_ENDPOINT: Value %!s(<nil>) not present
Key AZURE_SUBSCRIPTION_ID: Value %!s(<nil>) not present
Key ENABLE_MHC_CONTROL_PLANE: Value %!s(bool=true) not present
Key ENABLE_DEFAULT_STORAGE_CLASS: Value %!s(bool=true) not present
Key LDAP_USER_SEARCH_EMAIL_ATTRIBUTE: Value DN not present
Key OCI_IMAGE_ID: Value %!s(<nil>) not present
Key VSPHERE_TEMPLATE_MOID: Value %!s(<nil>) not present
Key NSXT_REMOTE_AUTH: Value false not present
Key ANTREA_NODEPORTLOCAL_ENABLED: Value %!s(bool=true) not present
Key AZURE_CLIENT_SECRET: Value %!s(<nil>) not present
Key AZURE_IMAGE_SKU: Value %!s(<nil>) not present
Key AUTOSCALER_MIN_SIZE_1: Value %!s(<nil>) not present
Key AVI_MANAGEMENT_CLUSTER_VIP_NETWORK_CIDR: Value %!s(<nil>) not present
Key ANTREA_NETWORKPOLICY_STATS: Value %!s(bool=false) not present
Key NSXT_ALLOW_UNVERIFIED_SSL: Value false not present
Key AWS_IDENTITY_REF_NAME: Value  not present
Key OCI_EXTERNAL_CONTROL_PLANE_EP_NSG_ID: Value %!s(<nil>) not present
Key AVI_CLOUD_NAME: Value %!s(<nil>) not present
Key OCI_PRIVATE_SERVICE_SUBNET_ID: Value %!s(<nil>) not present
Key AUTOSCALER_MAX_SIZE_2: Value %!s(<nil>) not present
Key AVI_CONTROL_PLANE_NETWORK: Value %!s(<nil>) not present
Key CONTROLPLANE_SIZE: Value %!s(<nil>) not present
Key VSPHERE_RESOURCE_POOL: Value %!s(<nil>) not present
Key AZURE_NODE_AZ: Value  not present
Key AZURE_CONTROL_PLANE_OUTBOUND_LB_FRONTEND_IP_COUNT: Value %!s(int=1) not present
Key AZURE_IDENTITY_NAMESPACE: Value %!s(<nil>) not present
Key AUTOSCALER_SCALE_DOWN_DELAY_AFTER_FAILURE: Value 3m not present
Key ANTREA_WIREGUARD_PORT: Value %!s(int=51820) not present
Key ANTREA_PROXY: Value %!s(bool=true) not present
Key OCI_COMPARTMENT_ID: Value %!s(<nil>) not present
Key AZURE_IMAGE_SUBSCRIPTION_ID: Value %!s(<nil>) not present
Key AVI_DATA_NETWORK: Value %!s(<nil>) not present
Key NSXT_POD_ROUTING_ENABLED: Value %!s(bool=false) not present
Key AZURE_VNET_CIDR: Value 10.0.0.0/16 not present
Key AZURE_ENABLE_PRIVATE_CLUSTER: Value %!s(bool=false) not present
Key OCI_EXTERNAL_CONTROL_PLANE_SUBNET_ID: Value %!s(<nil>) not present
Key AZURE_IMAGE_VERSION: Value %!s(<nil>) not present
Key ANTREA_EGRESS_EXCEPT_CIDRS: Value  not present
Key AZURE_CONTROL_PLANE_SUBNET_SECURITY_GROUP: Value  not present
Key AZURE_IMAGE_ID: Value %!s(<nil>) not present
Key TKG_CUSTOM_IMAGE_REPOSITORY_CA_CERTIFICATE: Value  not present
Key OIDC_DEX_CA: Value %!s(<nil>) not present
Key WORKER_MACHINE_COUNT_2: Value %!s(<nil>) not present
Key VSPHERE_SSH_AUTHORIZED_KEY: Value %!s(<nil>) not present
Key NSXT_VMC_AUTH_HOST: Value  not present
Key NSXT_SECRET_NAMESPACE: Value kube-system not present
Key AZURE_TENANT_ID: Value %!s(<nil>) not present
Key AZURE_CONTROL_PLANE_OS_DISK_STORAGE_ACCOUNT_TYPE: Value Premium_LRS not present
Key AZURE_IMAGE_NAME: Value %!s(<nil>) not present
Key ANTREA_ENABLE_BRIDGING_MODE: Value %!s(bool=false) not present
Key OCI_EXTERNAL_VCN_ID: Value %!s(<nil>) not present
Key VSPHERE_TEMPLATE: Value %!s(<nil>) not present
Key VSPHERE_AZ_1: Value %!s(<nil>) not present
Key AWS_LOAD_BALANCER_SCHEME_INTERNAL: Value %!s(bool=false) not present
Key NTP_SERVERS: Value %!s(<nil>) not present
Key AWS_SECURITY_GROUP_LB: Value  not present
Key AZURE_NODE_SUBNET_NAME: Value  not present
Key AZURE_ENABLE_NODE_DATA_DISK: Value %!s(bool=false) not present
Key AVI_CONTROLLER: Value %!s(<nil>) not present
Key AVI_PASSWORD: Value  not present
Key VSPHERE_REGION: Value %!s(<nil>) not present
Key CERT_DURATION: Value 2160h not present
Key AVI_CA_NAME: Value avi-controller-ca not present
Key OCI_NODE_MACHINE_TYPE_OCPUS: Value 1 not present
Key CLUSTER_API_SERVER_PORT: Value %!s(<nil>) not present
Key VIP_NETWORK_INTERFACE: Value eth0 not present
Key NSXT_USERNAME: Value  not present
Key AZURE_NODE_OS_DISK_STORAGE_ACCOUNT_TYPE: Value Premium_LRS not present
Key AUTOSCALER_MAX_NODES_TOTAL: Value 0 not present
Key AZURE_NODE_AZ_2: Value  not present
Key AZURE_CUSTOM_TAGS: Value %!s(<nil>) not present
Key AUTOSCALER_MAX_SIZE_0: Value %!s(<nil>) not present
Key VSPHERE_VERSION: Value %!s(<nil>) not present
Key boms: Value [map[bom_data:<nil> bom_name:<nil>]] not present
Key NSXT_SECRET_NAME: Value cloud-provider-vsphere-nsxt-credentials not present
Key AVI_MANAGEMENT_CLUSTER_VIP_NETWORK_NAME: Value %!s(<nil>) not present
Key ANTREA_TRAFFIC_ENCAP_MODE: Value encap not present
Key ANTREA_PROXY_LOAD_BALANCER_IPS: Value %!s(bool=false) not present
Key AZURE_ENABLE_ACCELERATED_NETWORKING: Value %!s(bool=true) not present
Key ANTREA_FLOWEXPORTER_COLLECTOR_ADDRESS: Value flow-aggregator.flow-aggregator.svc:4739:tls not present
Key ANTREA_SERVICE_EXTERNALIP: Value %!s(bool=false) not present
Key AUTOSCALER_MAX_SIZE_1: Value %!s(<nil>) not present
Key ANTREA_KUBE_APISERVER_OVERRIDE: Value %!s(<nil>) not present
Key WORKER_SIZE: Value %!s(<nil>) not present
Key VSPHERE_WORKER_MEM_MIB: Value %!s(int=4096) not present
Key IS_WINDOWS_WORKLOAD_CLUSTER: Value %!s(bool=false) not present
Key AZURE_NODE_SUBNET_SECURITY_GROUP: Value  not present
Key ANTREA_FLOWEXPORTER_POLL_INTERVAL: Value 5s not present
Key AWS_SECURITY_GROUP_CONTROLPLANE: Value  not present
Key NODE_MACHINE_TYPE_2: Value  not present
Key AZURE_CLIENT_ID: Value %!s(<nil>) not present
Key AZURE_NODE_OS_DISK_SIZE_GIB: Value %!s(int=128) not present
Key ENABLE_OIDC: Value %!s(bool=false) not present
Key DOCKER_MACHINE_TEMPLATE_IMAGE: Value %!s(<nil>) not present
Key ANTREA_IPAM: Value %!s(bool=false) not present
Key VSPHERE_WORKER_DISK_GIB: Value %!s(int=40) not present
Key VSPHERE_ZONE: Value %!s(<nil>) not present
Key CONTROL_PLANE_VM_CLASS: Value %!s(<nil>) not present
Key AZURE_NODE_OUTBOUND_LB_IDLE_TIMEOUT_IN_MINUTES: Value %!s(int=4) not present
Key MHC_FALSE_STATUS_TIMEOUT: Value 12m not present
Key ANTREA_EGRESS: Value %!s(bool=true) not present
Key CONTROL_PLANE_STORAGE_CLASS: Value %!s(<nil>) not present
Key WORKER_STORAGE_CLASS: Value %!s(<nil>) not present
Key NODE_POOL_0_LABELS: Value %!s(<nil>) not present
Key ANTREA_TRANSPORT_INTERFACE_CIDRS: Value  not present
Key ANTREA_DNS_SERVER_OVERRIDE: Value  not present
Key ANTREA_MULTICLUSTER_ENABLE: Value %!s(bool=false) not present
Key VSPHERE_WINDOWS_TEMPLATE: Value  not present
Key AZURE_ENABLE_CONTROL_PLANE_OUTBOUND_LB: Value %!s(bool=false) not present
Key AZURE_ENABLE_NODE_OUTBOUND_LB: Value %!s(bool=false) not present
Key MHC_UNKNOWN_STATUS_TIMEOUT: Value 5m not present
Key ANTREA_DISABLE_UDP_TUNNEL_OFFLOAD: Value %!s(bool=false) not present
Key NODE_STARTUP_TIMEOUT: Value 20m not present
Key OCI_CONTROL_PLANE_MACHINE_TYPE_OCPUS: Value 1 not present
Key NODE_POOL_0_NAME: Value workers not present
Key AZURE_LOCATION: Value %!s(<nil>) not present
Key AVI_INGRESS_SHARD_VS_SIZE: Value  not present
Key ANTREA_MULTICAST: Value %!s(bool=false) not present
Key VSPHERE_PASSWORD: Value %!s(<nil>) not present
Key WORKER_NODE_NAMESERVERS: Value %!s(<nil>) not present
Key OCI_EXTERNAL_WORKER_NSG_ID: Value %!s(<nil>) not present
Key OCI_EXTERNAL_CONTROL_PLANE_NSG_ID: Value %!s(<nil>) not present
Key AZURE_IMAGE_OFFER: Value %!s(<nil>) not present
Key OIDC_ISSUER_URL: Value %!s(<nil>) not present
Key ANTREA_FLOWEXPORTER_ACTIVE_TIMEOUT: Value 30s not present
Key ANTREA_TUNNEL_TYPE: Value geneve not present
Key ANTREA_TRACEFLOW: Value %!s(bool=true) not present
Key VSPHERE_WORKER_NUM_CPUS: Value %!s(int=2) not present
Key NSXT_ROUTER_PATH: Value  not present
Key AUTOSCALER_SCALE_DOWN_DELAY_AFTER_ADD: Value 10m not present
Key AUTOSCALER_SCALE_DOWN_DELAY_AFTER_DELETE: Value 10s not present
Key ANTREA_PROXY_NODEPORT_ADDRS: Value  not present
Key AZURE_RESOURCE_GROUP: Value  not present
Key AZURE_CONTROL_PLANE_SUBNET_NAME: Value  not present
Key SECRETGEN_CONTROLLER_ENABLE: Value %!s(bool=true) not present
Key AVI_NSXT_T1LR: Value %!s(<nil>) not present
Key DEPLOY_TKG_ON_VSPHERE7: Value %!s(<nil>) not present
Key AWS_SECURITY_GROUP_APISERVER_LB: Value  not present
Key AZURE_FRONTEND_PRIVATE_IP: Value 10.0.0.100 not present
Key OIDC_USERNAME_CLAIM: Value %!s(<nil>) not present
Key ANTREA_NODEPORTLOCAL_PORTRANGE: Value 61000-62000 not present
Key AZURE_IMAGE_RESOURCE_GROUP: Value %!s(<nil>) not present
Key CONTROL_PLANE_NODE_SEARCH_DOMAINS: Value %!s(<nil>) not present
Key FILTER_BY_ADDON_TYPE: Value %!s(<nil>) not present
Key AVI_ADMIN_CREDENTIAL_NAME: Value avi-controller-credentials not present
Key VSPHERE_NUM_CPUS: Value %!s(int=2) not present
Key VSPHERE_DATACENTER: Value %!s(<nil>) not present
Key VSPHERE_WORKER_CUSTOM_VMX_KEYS: Value %!s(<nil>) not present
Key AVI_MANAGEMENT_CLUSTER_CONTROL_PLANE_VIP_NETWORK_CIDR: Value %!s(<nil>) not present
Key VSPHERE_CONTROL_PLANE_DISK_GIB: Value %!s(int=40) not present
Key VSPHERE_CLONE_MODE: Value fullClone not present
Key NSXT_ROOT_CA_DATA_B64: Value  not present
Key AZURE_CONTROL_PLANE_SUBNET_CIDR: Value 10.0.0.0/24 not present
Key VSPHERE_SERVER: Value %!s(<nil>) not present
Key AZURE_SSH_PUBLIC_KEY_B64: Value %!s(<nil>) not present
Key AVI_MANAGEMENT_CLUSTER_SERVICE_ENGINE_GROUP: Value %!s(<nil>) not present
Key VSPHERE_CONTROL_PLANE_MEM_MIB: Value %!s(int=8192) not present
Key VSPHERE_USERNAME: Value %!s(<nil>) not present
Key VSPHERE_CONTROL_PLANE_PCI_DEVICES: Value %!s(<nil>) not present
Key AZURE_CONTROL_PLANE_DATA_DISK_SIZE_GIB: Value %!s(int=256) not present
Key AZURE_IMAGE_PUBLISHER: Value %!s(<nil>) not present
Key TKG_CUSTOM_IMAGE_REPOSITORY: Value  not present
```

Reply:
Thanks for clarification. Yes, TKGConfigReaderWriter only holds user's customized variables. Why do you need those default values only during applying to the management cluster? Shouldn't we use the same TKGConfigReaderWriter to interpolate ytt variables in clusterbootstrap.yaml for both bootstrap and management cluster?

Reply:
Got it. `TemplateDefinition` also includes `config_default.yaml` so it is also rendered with yttcb. I prefer to the later approach(Add path for `config_default.yaml`) because it is consistent with how we render them to the bootstrap cluster.